baseband/guppi/tests/test_guppi.py

```diff
@@ -0,0 +1,544 @@
+        assert frame.ndim == payload.ndim
+        assert np.all(frame[1337:1340] == np.array(
+            [[[11.+18.j, 4.+22.j, -9.-9.j, 15.-7.j],
+              [-10.-23.j, 15.-4.j, -21.+8.j, 11.-8.j]],
+             [[-6.+18.j, -13.+2.j, 5.-4.j, 36.-25.j],
+              [17.-27.j, 7.+7.j, -14.+4.j, 6.+5.j]],
+             [[-3.+8.j, 13.-25.j, 12.-18.j, 15.+7.j],
+              [26.-12.j, 18.-6.j, 6.+6.j, 6.+2.j]]], dtype=np.complex64))
+
+        with open(str(tmpdir.join('testguppi.raw')), 'w+b') as s:
+            frame.tofile(s)
+            s.seek(0)
+            frame2 = guppi.GUPPIFrame.fromfile(s, memmap=False)
+        assert frame2 == frame
+        frame3 = guppi.GUPPIFrame.fromdata(payload.data, header)
+        assert frame3 == frame
+        frame4 = guppi.GUPPIFrame.fromdata(payload.data, **header)
+        assert frame4 == frame
+        header5 = header.copy()
+        frame5 = guppi.GUPPIFrame(header5, payload, valid=False)
+        assert frame5.valid is False
+        assert np.all(frame5.data == 0.)
+        invalid_samples = frame5[-1000:]
+        assert np.all(invalid_samples == 0.)
+        assert invalid_samples.shape == (1000, 2, 4)
+        invalid_samples = frame5[8192:]
+        assert invalid_samples.shape == (0, 2, 4)
+        frame5.valid = True
+        assert frame5 == frame
+
+    def test_frame_memmap(self, tmpdir):
+        # Get frame regular way.
+        with guppi.open(SAMPLE_FILE, 'rb') as fr:
+            frame = fr.read_frame(memmap=False)
+        assert not isinstance(frame.payload.words, np.memmap)
+        # Check that if we map it instead, we get the same result.
+        with guppi.open(SAMPLE_FILE, 'rb') as fh:
+            frame2 = fh.read_frame(memmap=True)
+        assert frame2 == frame
+        assert isinstance(frame2.payload.words, np.memmap)
+        # Bit superfluous perhaps, but check decoding as well.
+        assert np.all(frame2[1337:1340] == np.array(
+            [[[11.+18.j, 4.+22.j, -9.-9.j, 15.-7.j],
+              [-10.-23.j, 15.-4.j, -21.+8.j, 11.-8.j]],
+             [[-6.+18.j, -13.+2.j, 5.-4.j, 36.-25.j],
+              [17.-27.j, 7.+7.j, -14.+4.j, 6.+5.j]],
+             [[-3.+8.j, 13.-25.j, 12.-18.j, 15.+7.j],
+              [26.-12.j, 18.-6.j, 6.+6.j, 6.+2.j]]], dtype=np.complex64))
+        assert np.all(frame2.data == frame.data)
+
+        # Now check writing.  First, without memmap, just ensuring writing
+        # to file works as well as to BytesIO done above.
+        filename = str(tmpdir.join('testguppi.raw'))
+        with guppi.open(filename, 'wb') as fw:
+            fw.write_frame(frame)
+
+        with guppi.open(filename, 'rb') as fw:
+            frame3 = fw.read_frame()
+
+        assert frame3 == frame
+        # Now memmap file to be written to.
+        with guppi.open(filename, 'wb') as fw:
+            frame4 = fw.memmap_frame(frame.header)
+        # Initially no data set, so frames should not match yet.
+        assert frame4 != frame
```

Comment:
Ahh I see what you mean - this is because `frame4` still exists in memory.  The same issue exists for DADA's test.  I'll change DADA's in the other PR.

Reply:
If you think this should be changed for both DADA and GUPPI, maybe we should make a mini-change after this PR?

Reply:
Sure, fine to do it separately.